xds/internal/xdsclient/creds/tls.go

```diff
@@ -0,0 +1,179 @@
+	cfg := &struct {
+		CertificateFile   string `json:"certificate_file"`
+		CACertificateFile string `json:"ca_certificate_file"`
+		PrivateKeyFile    string `json:"private_key_file"`
+	}{}
+
+	tlsConfig := tls.Config{}
+	if err := json.Unmarshal(jd, cfg); err != nil {
+		return nil, err
+	}
+
+	// We cannot simply always use a file_watcher provider because it behaves
+	// slightly differently from the xDS TLS config. Quoting A65:
+	//
+	// > The only difference between the file-watcher certificate provider
+	// > config and this one is that in the file-watcher certificate provider,
+	// > at least one of the "certificate_file" or "ca_certificate_file" fields
+	// > must be specified, whereas in this configuration, it is acceptable to
+	// > specify neither one.
+	//
+	// We only use a file_watcher provider if either one of them or both are
+	// specified.
+	if cfg.CACertificateFile != "" || cfg.CertificateFile != "" || cfg.PrivateKeyFile != "" {
+		// file_watcher currently ignores BuildOptions.
+		provider, err := certprovider.GetProvider("file_watcher", jd, certprovider.BuildOptions{})
+		if err != nil {
+			// GetProvider fails if jd is invalid, e.g. if only one of private
+			// key and certificate is specified.
+			return nil, fmt.Errorf("failed to get TLS provider: %w", err)
+		}
+		if cfg.CertificateFile != "" {
+			tlsConfig.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				// Client cert reloading for mTLS.
+				km, err := provider.KeyMaterial(context.Background())
+				if err != nil {
+					return nil, err
+				}
+				if len(km.Certs) != 1 {
+					// xDS bootstrap has a single private key file, so there
+					// must be exactly one certificate or certificate chain
+					// matching this private key.
+					return nil, fmt.Errorf("certificate_file must contains exactly one certificate or certificate chain")
+				}
+				return &km.Certs[0], nil
+			}
+			if cfg.CACertificateFile == "" {
+				// No need for a callback to load the CA. Use the normal mTLS
+				// transport credentials.
+				return &tlsBundle{
+					jd:                   jd,
+					transportCredentials: credentials.NewTLS(&tlsConfig),
+				}, nil
+			}
+		}
+		return &tlsBundle{
+			jd: jd,
+			transportCredentials: &caReloadingClientTLSCreds{
+				baseConfig: &tlsConfig,
+				provider:   provider,
+			},
+		}, nil
+	}
+
+	// None of certificate_file and ca_certificate_file are set.
+	// Use the system-wide root certs.
+	return &tlsBundle{
+		jd:                   jd,
+		transportCredentials: credentials.NewTLS(&tlsConfig),
+	}, nil
+}
+
+func (t *tlsBundle) TransportCredentials() credentials.TransportCredentials {
+	return t.transportCredentials
+}
+
+func (t *tlsBundle) PerRPCCredentials() credentials.PerRPCCredentials {
+	// mTLS provides transport credentials only. There are no per-RPC
+	// credentials.
+	return nil
+}
+
+func (t *tlsBundle) NewWithMode(_ string) (credentials.Bundle, error) {
+	return NewTLS(t.jd)
```

Comment:
I took your suggestion, it makes sense, we shouldn't switch mode with this bundle.

I struggled to understand why I need to implement this in the first place. I was a bit confused with what to do with the "mode" mechanism, but looking into it a bit more it (kind of) makes sense. I see it's used for rls and grpclb in order to change the credentials configuration after it was instantiated and passed through a dial option by the user -- IIUC this is so that the "google" cred bundle can embed a different configuration for the control plane (RLS or grpclb) -- code here: https://github.com/grpc/grpc-go/blob/9353ae3bb432cf1075bf9329ab264421b8e01394/credentials/google/google.go#L128-L138.
 So `NewWithMode` allows changing the credentials within gRPC based on the context they are used in (e.g. a data plane endpoint or a control plane endpoint) *without having separate bundles*.